### PR TITLE
AEA-1554 Script to detect if aea and plugins need updating

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Add script to detect if aea and plugins need updating.

## Fixes

n/a

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: scripts/bump_aea_version.py [hunk 1]
```diff
@@ -18,178 +18,467 @@
...
-    DEFAULT_CONTRACT_CONFIG_FILE,
-    DEFAULT_PROTOCOL_CONFIG_FILE,
-    DEFAULT_SKILL_CONFIG_FILE,
-)
 from aea.helpers.base import compute_specifier_from_version
 from scripts.generate_ipfs_hashes import update_hashes
 
 
-VERSION_NUMBER_PART_REGEX = r"(0|[1-9]\d*)"
-VERSION_REGEX = fr"(any|latest|({VERSION_NUMBER_PART_REGEX})\.({VERSION_NUMBER_PART_REGEX})\.({VERSION_NUMBER_PART_REGEX})(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)"
```
Comment (marcofavorito) [inline scripts/bump_aea_version.py:42]: unused

File: scripts/bump_aea_version.py [hunk 2]
```diff
@@ -18,178 +18,467 @@
...
+in JSON format.
 
+e.g.:
+"aea-ledger-fetchai": {"version": ">=2.0.0, <3.0.0"}
+"""
+JSON_DEPENDENCY_SPECIFIER_SET_PATTERN = (
+    '(?<="{package_name}": ."version": ")({specifier_set})(?=".)'
+)
 
-def update_version_for_files(current_version: str, new_version: str) -> None:
```
Comment (marcofavorito) [inline scripts/bump_aea_version.py:62]: the additions below are mostly a refactoring of the old implementation. Unfortunately git changes are not very helpful

File: scripts/bump_aea_version.py [hunk 3]
```diff
@@ -18,178 +18,467 @@
...
-    with open(path, "rt") as fin:
-        for line in fin:
-            if "__version__" not in line:
-                continue
-            match = re.search('__version__ = "(.*)"', line)
-            if match is None:
-                raise ValueError("Current version is not well formatted.")
-            current_version = match.group(1)
-    if current_version == "":
-        raise ValueError("No version found!")
```
Comment (marcofavorito) [inline scripts/bump_aea_version.py:102]: this only checked for `__version__`; now we also check `setup.py` if the former is not present

File: scripts/bump_aea_version.py [hunk 4]
```diff
@@ -18,178 +18,467 @@
...
 
-def update_version_for_file(path: Path, current_version: str, new_version: str) -> None:
-    """
-    Update version for file.
+    @wraps(func)
+    def wrapper(self: Any, *args: Any, **kwargs: Any) -> None:
+        if self.is_executed:
+            raise ValueError("already executed")
+        self._executed = True
+        self._result = func(self, *args, **kwargs)
```
Comment (marcofavorito) [inline scripts/bump_aea_version.py:123]: just used to implement the functor pattern

**Comment by marcofavorito** [comment]
Try it with:

```
scripts/bump_aea_version.py --new-version 3.0.0 -p aea-ledger-fetchai=2.0.1
```

It should print something like:
```
[2021-05-11 14:46:53,211][root][INFO] Parsed arguments: Namespace(new_version='3.0.0', no_fingerprint=False, plugin_new_version=['aea-ledger-fetchai=2.0.1'])
[2021-05-11 14:46:53,211][root][INFO] Parsed plugin versions: {'aea-ledger-fetchai': <Version('2.0.1')>}
[2021-05-11 14:46:53,235][root][INFO] Running 'git diff with args: ('v1.0.1', '--', 'aea')'
[2021-05-11 14:46:53,243][root][INFO] The package aea has no changes since last tag.
[2021-05-11 14:46:53,243][root][INFO] AEA package processed.
[2021-05-11 14:46:53,243][root][INFO] Processing plugins:
[2021-05-11 14:46:53,243][root][INFO] Skipping aea-cli-ipfs as it is not specified in input {'aea-ledger-fetchai': <Version('2.0.1')>}
[2021-05-11 14:46:53,243][root][INFO] Skipping aea-ledger-ethereum as it is not specified in input {'aea-ledger-fetchai': <Version('2.0.1')>}
[2021-05-11 14:46:53,243][root][INFO] Skipping aea-ledger-cosmos as it is not specified in input {'aea-ledger-fetchai': <Version('2.0.1')>}
[2021-05-11 14:46:53,243][root][INFO] Processing aea-ledger-fetchai: upgrading at version 2.0.1
[2021-05-11 14:46:53,247][root][INFO] Running 'git diff with args: ('v1.0.1', '--', 'plugins/aea-ledger-fetchai/aea_ledger_fetchai')'
[2021-05-11 14:46:53,253][root][INFO] The package plugins/aea-ledger-fetchai/aea_ledger_fetchai has no changes since last tag.
[2021-05-11 14:46:53,253][root][INFO] OK
[2021-05-11 14:46:53,253][root][INFO] Not updating fingerprint, since no specifier set has been updated.
```

File: scripts/bump_aea_version.py [hunk 5]
```diff
@@ -18,178 +18,465 @@
...
+        logging.info("Not updating fingerprint, since --no-fingerprint was specified.")
+    elif have_updated_specifier_set is False:
+        logging.info(
+            "Not updating fingerprint, since no specifier set has been updated."
+        )
     else:
-        print("Updating hashes and fingerprints.")
+        logging.info("Updating hashes and fingerprint.")
         return_code = update_hashes()
     sys.exit(return_code)
```
Comment (marcofavorito) [inline scripts/bump_aea_version.py:482]: pseudocode:

- parse parameters
- if repo dirty, exit with error
- bump aea version
- bump all plugins
- if *only* specifier sets are updated, recompute AEA package hashes

File: scripts/bump_aea_version.py [hunk 6]
```diff
@@ -18,178 +18,465 @@
...
+        alternatives.append(f"{lower} *, *{upper}")
+        return "|".join(alternatives)
+
+    def is_different_from_latest_tag(self) -> bool:
+        """Check whether the package has changes since the latest tag."""
+        assert len(self.repo.tags) > 0, "no git tags found"
+        latest_tag_str = str(self.repo.tags[-1])
+        args = latest_tag_str, "--", str(self.python_pkg_dir)
+        logging.info(f"Running 'git diff with args: {args}'")
+        diff = self.repo.git.diff(*args)
```
[the remark this answers is not in the document]
Comment (marcofavorito) [reply scripts/bump_aea_version.py:375]: command template: 
```
git diff {tag} -- {subdirectory-to-check}
```

**Review by solarw — APPROVED** [review]
LGTM

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**